src/Authenticator/OauthAuthenticator.php

```diff
@@ -0,0 +1,243 @@
+
+        if (empty($this->_provider) || !array_key_exists($this->_provider, $this->_authService->getConfig()['provider'])) {
+            return false;
+        }
+
+        if (!empty($redirectUrl)) {
+            if (is_array($redirectUrl)) {
+                $redirectUrl = Router::url($redirectUrl);
+            }
+            $redirectUrl = $redirectUrl . '/' . $this->_provider . '/';
+
+            return strcasecmp($request->getUri()->getPath(), $redirectUrl) === 0;
+        }
+
+        return false;
+    }
+
+    /**
+     * Checks the requests if it is the configured login action
+     *
+     * @param \Cake\Http\ServerRequest $request The current request.
+     * @return bool
+     */
+    protected function _isLoginUrl(ServerRequest $request)
+    {
+        $loginUrl = $this->getConfig('loginUrl');
+        $this->_provider = $request->getQuery('provider');
+
+        if (empty($this->_provider) || !array_key_exists($this->_provider, $this->_authService->getConfig()['provider'])) {
+            return false;
+        }
+
+        if (!empty($loginUrl)) {
+            if (is_array($loginUrl)) {
+                $loginUrl = Router::url($loginUrl);
+            }
+
+            return strcasecmp($request->getUri()->getPath(), $loginUrl) === 0;
+        }
+
+        return false;
+    }
+
+    /**
+     * Check the fields to ensure they are supplied
+     *
+     * @param array $identity The identity returned by the provider.
+     * @return bool False if the fields have not been supplied. True if they exist.
+     */
+    protected function _checkOauthIdentity($identity)
+    {
+        foreach ($this->getConfig('fields') as $field) {
+            if (!isset($identity[$field])) {
+                return false;
+            }
+
+            $value = $identity[$field];
+            if (empty($value) || !is_string($value)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Fetch the identity from the provider
+     *
+     * @param string $queryParameters Query Parameter
+     * @return \SocialConnect\Common\Entity\User User entity
+     */
+    protected function _getIdentity($queryParameters)
+    {
+        $provider = $this->_authService->getProvider($this->_provider);
+        $accessToken = $provider->getAccessTokenByRequestParameters($queryParameters);
+
+        return $provider->getIdentity($accessToken);
+    }
+
+    /**
+     * Redirect the user to the provider
+     *
+     * @param \Cake\Http\Response $response Response object.
+     * @return void
+     */
+    protected function _redirect(Response $response)
+    {
+        $provider = $this->_authService->getProvider($this->_provider);
+
+        $response = $response
+            ->withStatus(302)
+            ->withLocation($provider->makeAuthUrl());
+
+        $emitter = new ResponseEmitter();
+        $emitter->emit($response);
+
+        exit;
```

Comment:
This would require to check the result of `authenticate` and redirect in the `AuthenticationMiddleware` if a response object is returned?
Don't know if this is a good idea either 🤔
I'm totally with you that this is a "dirty" fix, but I haven't found any other PSR7 conform solution to redirect early.